Objects/regions.c

```diff
@@ -92,9 +116,324 @@ struct regionmetadata {
     regionmetadata* next;
 };
 
-static bool is_bridge_object(PyObject *op);
-static int regionmetadata_has_ancestor(regionmetadata* data, regionmetadata* other);
-static regionmetadata* PyRegion_get_metadata(PyRegionObject* obj);
+static Py_region_ptr_t regionmetadata_get_root(Py_region_ptr_t self)
+{
+    // Test for constants
+    if (IS_REGION_NULL(self) || IS_CONST_REGION(self)) {
+        return _Py_CAST(Py_region_ptr_t, self);
+    }
+
+    // Return self if it wasn't merged with another region
+    regionmetadata* self_data = _Py_CAST(regionmetadata*, self);
+    if (!REGION_PTR_GET_TAG(self_data->parent, Py_METADATA_MERGE_TAG)) {
+        return self;
+    }
+
+    // FIXME: It can happen that there are several layers in this union-find
+    // structure. It would be efficient to directly update the parent pointers
+    // for deeper nodes.
+    return regionmetadata_get_root(Py_region_ptr(self_data->parent));
+}
+#define regionmetadata_get_root(self) \
+    regionmetadata_get_root(_Py_CAST(Py_region_ptr_t, self))
+
+// This returns the represenative of this union-find structure or NULL
+// if the representative is a constant value.
+static regionmetadata* regionmetadata_get_root_or_null(Py_region_ptr_t self)
+{
+    Py_region_ptr_t rep = regionmetadata_get_root(self);
+    // Test for constants
+    if (IS_CONST_REGION(rep)) {
+        return NULL;
+    }
+
+    return _Py_CAST(regionmetadata*, rep);
+}
+#define regionmetadata_get_root_or_null(self) \
+    regionmetadata_get_root_or_null(_Py_CAST(Py_region_ptr_t, (self)))
+
+static void regionmetadata_open(regionmetadata* self) {
+    self = regionmetadata_get_root_or_null(self);
+    if (self) {
+        self->is_open = true;
+    }
+}
+
+static bool regionmetadata_is_open(regionmetadata* self) {
+    self = regionmetadata_get_root_or_null(self);
+    if (!self) {
+        // All constant regions are open
+        return true;
+    }
+
+    return self->is_open;
+}
+
+static void regionmetadata_inc_rc(Py_region_ptr_t self_raw)
+{
+    // Constant regions don't keep track of open subregions.
+    if (IS_REGION_NULL(self_raw) || IS_CONST_REGION(self_raw)) {
+        return;
+    }
+
+    regionmetadata* self = _Py_CAST(regionmetadata*, self_raw);
+
+    // This doesn't use the representative, since we want
```

Comment:
Above you use the pattern
```
self = regionmetadata_get_root_or_null(self);
```
Maybe that should be rewritten as
```
regionmetadata* representative = regionmetadata_get_root_or_null(self);
```
To more clearly distinguish between cases where we do need to get the representative rather than the "self". Just a thought.